import UserFront from "../models/UserFront"
import UserData from "../models/UserData"
import UserUpdate from "../models/UserUpdate"
import UserCreate from "../models/UserCreate"
import NewPassword from "../models/NewPassword"
import Email from "../models/Email"

// Login
export const userFetch = async(userFront: UserFront)=>{
    try{
        //Configuraci칩n de la solicitud
        const response = await fetch("http://localhost:8080/auth/login",{ 
            method: "POST",
            headers: {"Content-Type": "application/json"},
            body: JSON.stringify(userFront)
        })
        if(response.ok){
            return await response.json()
        }else{
            throw new Error(`Los datos ingresados son incorrectos. Intentalo de nuevo.`);
        }
    }catch(e){
        if (e instanceof Error) {
            if (e.message === "Failed to fetch") {
                throw new Error("No se pudo conectar al servidor. Vuelve a intentarlo m치s tarde.");
            } else {            
            throw new Error(e.message);
            }
        }
        throw new Error('Error desconocido');         
    }
}
//ChangeCode
export async function modifyPassword(body: NewPassword){
    try{
        const response = await fetch("http://localhost:8080/auth/reset",{  
            method: "POST",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify(body)
        });
        if(response.ok){
            const data = await response.json();
            return data;
        }
        throw new Error("No se pudo modifcar la contrase침a");  
        
    }catch(e){
        if (e instanceof Error) {
            throw new Error(e.message);
          }
          throw new Error('Error desconocido'); 
    }     
}
//PasswordRecovery
export async function passRecovery(body: Email ){
    try{
        const response = await fetch("http://localhost:8080/auth/forgot",{  
            method: "POST",
            headers: {
              "Content-Type": "application/json",
            },
            body: JSON.stringify(body)
        });
        if(response.ok){
            const data = await response.json();
            return data;
        }
        throw new Error("No se pudo modifcar la contrase침a");  
        
    }catch(e){
        if (e instanceof Error) {
            throw new Error(e.message);
          }
          throw new Error('Error desconocido'); 
    }     
}
// User Management
export async function getUsers(token :string):Promise<UserData[]>{
    try{
        const response = await fetch("http://localhost:8080/users",{  
            method: "GET",
            headers: {
              'Authorization': `Bearer ${token}`,
            },
        });
        if(response.ok){
            const data: UserData[] = await response.json();
            return data;
        }
        throw new Error("No se pudo obtener los usuarios");  
    }catch(e){  
        if (e instanceof Error) {
            throw new Error(e.message);
          }
          throw new Error('Error desconocido'); 
    }
}
export async function updateUser(token: string, id_user: number, body: UserUpdate){
    try{
        const response = await fetch(`http://localhost:8080/users/update?id=${id_user}`,{  
            method: "PUT",
            headers: {
              "Content-Type": "application/json",
              'Authorization': `Bearer ${token}`,
            },
            body: JSON.stringify(body)
        });
        if(response.ok){
            const data = await response.json();
            return data;
        }
        throw new Error("No se pudo actualizar los datos el usuario");  
        
    }catch(e){
        if (e instanceof Error) {
            throw new Error(e.message);
          }
          throw new Error('Error desconocido'); 
    }     
}
export async function addUser(token: string, body: UserCreate){
    try{
        const response = await fetch(`http://localhost:8080/users/register`,{  
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              'Authorization': `Bearer ${token}`,
            },
            body: JSON.stringify(body)
        });
        if(response.ok){
            const data = await response.json();
            return data;
        }
        throw new Error("No se pudo agregar el usuario");  
        
    }catch(e){
        if (e instanceof Error) {
            throw new Error(e.message);
          }
          throw new Error('Error desconocido'); 
    }     
}
export async function stateUser(token: string, id_user: number, body: string){
    try{
        const response = await fetch(`http://localhost:8080/users/toggle-activation?id=${id_user}&status=${body}`,{  
            method: "PUT",
            headers: {
              "Content-Type": "application/json",
              'Authorization': `Bearer ${token}`,
            },
            body: JSON.stringify(body)
        });
        if(response.ok){
            const data = await response.json();
            return data;
        }
        throw new Error("No se pudo modificar el estado del usuario");  
        
    }catch(e){
        if (e instanceof Error) {
            throw new Error(e.message);
          }
          throw new Error('Error desconocido'); 
    }     
}
export async function searchUsers(token: string, searchData: string | number):Promise<UserData[]>{
    const url = typeof searchData === "string" ? "name" : "dni";
    try{
        const response = await fetch(`http://localhost:8080/users/search?${url}=${searchData}`,{  
            method: "GET",
            headers: {
              'Authorization': `Bearer ${token}`,
            },
        });
        if(response.ok){
            const data: UserData[] = await response.json();
            return data;
        }
        throw new Error("No se encontraron resultados");  
    }catch(e){  
        if (e instanceof Error) {
            throw new Error(e.message);
          }
          throw new Error('Error desconocido'); 
    }
}

